lxd/storage/drivers/drivers_lvm_utils.go

```diff
@@ -0,0 +1,693 @@
+	args := []string{"-n", snapLvName, "-s", srcVolDevPath}
+
+	if isRecent {
+		args = append(args, "-kn")
+	}
+
+	// If the source is not a thin volume the size needs to be specified.
+	// According to LVM tools 15-20% of the original volume should be sufficient.
+	// However, let's not be stingy at first otherwise we might force users to fiddle around with lvextend.
+	if !makeThinLv {
+		lvSizeBytes, err := d.roundedSizeBytesString(d.volumeSize(snapVol))
+		if err != nil {
+			return "", err
+		}
+
+		args = append(args, "--size", fmt.Sprintf("%db", lvSizeBytes))
+		logCtx["size"] = fmt.Sprintf("%db", lvSizeBytes)
+	}
+
+	if readonly {
+		args = append(args, "-pr")
+	} else {
+		args = append(args, "-prw")
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	_, err = shared.TryRunCommand("lvcreate", args...)
+	if err != nil {
+		return "", fmt.Errorf("Error creating LVM logical volume snapshot: %s to %s: %v", srcVol.name, snapVol.name, err)
+	}
+	d.logger.Debug("Logical volume snapshot created", logCtx)
+
+	revert.Add(func() {
+		d.removeLogicalVolume(d.lvmDevPath(vgName, snapVol.volType, snapVol.contentType, snapVol.name))
+	})
+
+	targetVolDevPath := d.lvmDevPath(vgName, snapVol.volType, snapVol.contentType, snapVol.name)
+	if makeThinLv {
+		// Snapshots of thin logical volumes can be directly activated.
+		// Normal snapshots will complain about changing the origin (Which they never do.),
+		// so skip the activation since the logical volume will be automatically activated anyway.
+		err := d.activeLogicalVolume(targetVolDevPath)
+		if err != nil {
+			return "", fmt.Errorf("Error activating LVM logical volume snapshot: %s: %v", snapVol.name, err)
+		}
+	}
+
+	revert.Success()
+	return targetVolDevPath, nil
+}
+
+// activeLogicalVolume marks the logical volume as active.
+func (d *lvm) activeLogicalVolume(volDevPath string) error {
+	_, err := shared.TryRunCommand("lvchange", "-ay", volDevPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// removeLogicalVolume removes a logical volume.
+func (d *lvm) removeLogicalVolume(volDevPath string) error {
+	_, err := shared.TryRunCommand("lvremove", "-f", volDevPath)
+	if err != nil {
+		return fmt.Errorf("Error removing LVM logical volume %s: %v", volDevPath, err)
+	}
+	d.logger.Debug("Logical volume removed", log.Ctx{"dev": volDevPath})
+
+	return nil
+}
+
+// renameLogicalVolume renames a logical volume.
+func (d *lvm) renameLogicalVolume(volDevPath string, newVolDevPath string) error {
+	_, err := shared.TryRunCommand("lvrename", volDevPath, newVolDevPath)
+	if err != nil {
+		return fmt.Errorf("Error renaming LVM logical volume from \"%s\" to \"%s\": %v", volDevPath, newVolDevPath, err)
+	}
+	d.logger.Debug("Logical volume renamed", log.Ctx{"dev": volDevPath, "new_dev": newVolDevPath})
+
+	return nil
+}
+
+// volumeTypeToLVMType converts volume type to internal volume prefix.
+func (d *lvm) volumeTypeToLVMType(volType VolumeType) (string, error) {
```

Comment:
Only has a single caller, any reason to keep it separate?

Reply:
@stgraber no will merge into lvmFullVolumeName